src/OrleansGoogleCloudPlatformUtils/ProtobufSerializer.cs

```diff
@@ -0,0 +1,166 @@
+                    var parser = itemType.GetProperty("Parser", BindingFlags.Public | BindingFlags.Static).GetValue(null, null);
+                    Parsers.TryAdd(itemType.TypeHandle, parser);
+                }
+                return true;
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Creates a deep copy of an object
+        /// </summary>
+        /// <param name="source">The source object to be copy</param>
+        /// <returns>The copy that was created</returns>
+        public object DeepCopy(object source)
+        {
+            if (source == null)
+            {
+                return null;
+            }
+
+            dynamic dynamicSource = source;
+            return dynamicSource.Clone();
+        }
+
+        /// <summary>
+        /// Serializes an object to a binary stream
+        /// </summary>
+        /// <param name="item">The object to serialize</param>
+        /// <param name="writer">The <see cref="BinaryTokenStreamWriter"/></param>
+        /// <param name="expectedType">The type the deserializer should expect</param>
+        public void Serialize(object item, BinaryTokenStreamWriter writer, Type expectedType)
+        {
+            if (writer == null)
+            {
+                throw new ArgumentNullException("writer");
+            }
+
+            if (item == null)
+            {
+                // Special handling for null value. 
+                // Since in this ProtobufSerializer we are usualy writing the data lengh as 4 bytes
+                // we also have to write the Null object as 4 bytes lengh of zero.
+                writer.Write(0);
+                return;
+            }
+
+            IMessage iMessage = item as IMessage;
+            if (iMessage == null)
+            {
+                throw new ArgumentException("The provided item for serialization in not an instance of " + typeof(IMessage), "item");
+            }
+            // The way we write the data is potentially in-efficinet, 
+            // since we are first writing to ProtoBuff's internal CodedOutputStream
+            // and then take its internal byte[] and write it into out own BinaryTokenStreamWriter.
+            // Writing byte[] to BinaryTokenStreamWriter may sometimes copy the byte[] and sometimes just append ass ArraySegment without copy.
+            // In the former case it will be a secodnd copy.
+            // It would be more effecient to write directly into BinaryTokenStreamWriter
+            // but protobuff does not currently support writing directly into a given arbitary stream
+            // (it does support System.IO.Steam but BinaryTokenStreamWriter is not compatible with System.IO.Steam).
+            // Alternatively, we could force to always append to BinaryTokenStreamWriter, but that could create a lot of small ArraySegments.
+            // The plan is to ask the ProtoBuff team to add support for some "InputStream" interface, like Bond does.
```

Comment:
Yes, we can also do that. But that would mean we need to make our `BinaryTokenStreamWriter` a subclass of `System.IO.Stream`. Which may be a problem going forward, since it is base class we may not have the same flexibility to be in 100% control as we would have been with an interface (`System.IO.Stream` might control something that cannot be overwritten by a derived class).
I was mostly thinking about the idea of a simper intermediate InputStream interface similar to the one Bond has.

Anyway, this is a future optimization and I will follow up with protobuf team (@jskeet) after this is merged.